ingress-watcher/cmd/push.go

```diff
@@ -0,0 +1,90 @@
+package cmd
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"time"
+
+	"github.com/cybozu-go/log"
+	"github.com/cybozu-go/well"
+	"github.com/cybozu/neco-containers/ingress-watcher/pkg/watch"
+	"github.com/prometheus/client_golang/prometheus/push"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+var pushConfig struct {
+	JobName      string
+	PushAddr     string
+	PushInterval time.Duration
+}
+
+var pushCmd = &cobra.Command{
+	Use:   "push",
+	Short: "Push metrics to Pushgateway",
+	Long:  `Push metrics to Pushgateway`,
+	PreRunE: func(cmd *cobra.Command, args []string) error {
+		if configFile != "" {
+			if err := viper.Unmarshal(&pushConfig); err != nil {
+				return err
+			}
+		}
+
+		if len(pushConfig.JobName) == 0 {
+			return errors.New("required flag \"job-name\" not set")
+		}
+
+		if len(pushConfig.PushAddr) == 0 {
+			return errors.New("required flag \"push-addr\" not set")
+		}
+
+		return nil
+	},
+	Run: func(cmd *cobra.Command, args []string) {
+		well.Go(watch.NewWatcher(
+			rootConfig.TargetURLs,
+			rootConfig.WatchInterval,
+			&http.Client{},
+		).Run)
+		well.Go(func(ctx context.Context) error {
+			tick := time.NewTicker(pushConfig.PushInterval)
+			defer tick.Stop()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case <-tick.C:
+					pusher := push.New(pushConfig.PushAddr, pushConfig.JobName).Gatherer(registry)
+					err := pusher.Add()
+					if err != nil {
```

Comment:
If there is an error, such as a failure to connect to pushgateway, can we notice the error?

Reply:
There is a metric that records the last pushed timestamp.
So, if the watcher continues to fail pushing the metrics, the timestamp gets old and we can notice that by alert.